nixos/modules/services/misc/matrix-conduit.nix

```diff
@@ -0,0 +1,142 @@
+            };
+            global.max_request_size = mkOption {
+              type = types.ints.positive;
+              default = 20000000;
+              description = "Max request size in bytes. Don't forget to also change it in the proxy.";
+            };
+            global.allow_registration = mkOption {
+              type = types.bool;
+              default = false;
+              description = "Whether new users can register on this server.";
+            };
+            global.allow_encryption = mkOption {
+              type = types.bool;
+              default = true;
+              description = "Whether new encrypted rooms can be created. Note: existing rooms will continue to work.";
+            };
+            global.allow_federation = mkOption {
+              type = types.bool;
+              default = true;
+              description = ''
+                Whether this server federates with other servers.
+              '';
+            };
+            global.trusted_servers = mkOption {
+              type = types.listOf types.str;
+              default = [ "matrix.org" ];
+              description = "Servers trusted with signing server keys.";
+            };
+
+            global.address = mkOption {
+              type = types.str;
+              default = "::1";
+              description = "Address to listen on for connections by the reverse proxy/tls terminator.";
+            };
+            global.database_path = mkOption {
+              type = types.str;
+              default = "/var/lib/matrix-conduit/";
+              readOnly = true;
+              description = ''
+                Path to the conduit database, the directory where conduit will save its data.
+                Note that due to using the DynamicUser feature of systemd, this value should not be changed
+                and is set to be read only.
+              '';
+            };
+          };
+        };
+        default = {};
+        description = ''
+            Generates the conduit.toml configuration file. Refer to
+            <link xlink:href="https://gitlab.com/famedly/conduit/-/blob/master/conduit-example.toml"/>
+            for details on supported values.
+            Note that database_path can not be edited because the service's reliance on systemd StateDir.
+        '';
+      };
+    };
+
+
+    config = mkIf cfg.enable {
+      systemd.services.conduit = {
+        description = "Conduit Matrix Server";
+        documentation = [ "https://gitlab.com/famedly/conduit/" ];
+        wantedBy = [ "multi-user.target" ];
+        environment = lib.mkMerge ([
+          { CONDUIT_CONFIG = configFile; }
+          cfg.extraEnvironment
+        ]);
+        serviceConfig = {
+          DynamicUser = true;
```

Comment:
don't know if we could restrict the permissions even further. Unfortunately upstream seems to have no recommendation.

Reply:
I think I've gone as far as possible. If you know something else to try out, let me know. Upstreams don't usually have much knowledge about Systemd harderning, unfortunately.

Reply:
I think everything except the system call filter (which doesn't work yet) could be done.
```diff
diff --git a/nixos/modules/services/misc/matrix-conduit.nix b/nixos/modules/services/misc/matrix-conduit.nix
index 3f62f3aea87..69622f0381e 100644
--- a/nixos/modules/services/misc/matrix-conduit.nix
+++ b/nixos/modules/services/misc/matrix-conduit.nix
@@ -129,24 +129,31 @@ in
           User = "conduit";
           LockPersonality = true;
           MemoryDenyWriteExecute = true;
+          CapabilityBoundingSet = "";
+          DevicePolicy = "strict";
+          DeviceAllow = [ "/dev/urandom" "/dev/random" "/dev/stdin" "/dev/stdout" "/dev/stderr" ];
           ProtectClock = true;
           ProtectControlGroups = true;
+          ProtectHome = true;
           ProtectHostname = true;
           ProtectKernelLogs = true;
           ProtectKernelModules = true;
           ProtectKernelTunables = true;
+          ProtectProc = "invisible";
           PrivateDevices = true;
           PrivateMounts = true;
           PrivateUsers = true;
+          ProcSubset = "pid";
           RestrictAddressFamilies = [ "AF_INET" "AF_INET6" ];
           RestrictNamespaces = true;
           RestrictRealtime = true;
           SystemCallArchitectures = "native";
           SystemCallFilter = [
-            "@system-service"
-            "~@privileged"
+            # inspiration: systemd-analyze syscall-filter
+            "@default @basic-io @chown @file-system @io-event @network-io @process @signal @sync mprotect"
           ];
           StateDirectory = "matrix-conduit";
+          UMask = "0077";
           ExecStart = "${cfg.package}/bin/conduit";
           Restart = "on-failure";
           RestartSec = 10;

```

Reply:
Thanks for the diff. I suggest we put this on the backlog and open that topic in a new PR once this one is merged.